homeassistant/components/recorder/__init__.py

```diff
@@ -195,31 +209,41 @@ def __init__(self, hass: HomeAssistant, purge_days: int, uri: str,
 
         def start_recording(event):
             """Start recording."""
-            self.start()
+            self.start_recording.set()
 
         hass.bus.listen_once(EVENT_HOMEASSISTANT_START, start_recording)
         hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)
         hass.bus.listen(MATCH_ALL, self.event_listener)
 
+        self.get_session = None
+
+        self.start()
+
     def run(self):
         """Start processing events to save."""
         from homeassistant.components.recorder.models import Events, States
-        import sqlalchemy.exc
+        from homeassistant.components.recorder.restore_state import (
+            load_restore_cache)
+        from sqlalchemy.exc import SQLAlchemyError
 
         while True:
             try:
                 self._setup_connection()
                 self._setup_run()
                 break
-            except sqlalchemy.exc.SQLAlchemyError as err:
+            except SQLAlchemyError as err:
                 _LOGGER.error("Error during connection setup: %s (retrying "
                               "in %s seconds)", err, CONNECT_RETRY_WAIT)
                 time.sleep(CONNECT_RETRY_WAIT)
 
+        load_restore_cache(self.hass)
```

Comment:
Can be done, I've got a new unit test (should be non-flakey) where I discovered I'll have to `yield get_last_state` in any case. On startup many components will potentially wait for this event, which hopefully won't affect order of adding to hass. At least by the time this is called "entities" was already updated in entity_component